### PR TITLE
Harden consolidation checks

### PR DESCRIPTION
**Pull request by gnusi**

### CODE REVIEW

File: core/index/index_writer.cpp [hunk 1]
```diff
@@ -1461,10 +1467,11 @@ index_writer::consolidation_result index_writer::consolidate(
     // check that candidates are not involved in ongoing merges
     for (const auto* candidate : candidates) {
       // segment has been already chosen for consolidation (or at least was choosen), give up
-      if (consolidating_segments_.contains(candidate)) {
+      if (!candidate || consolidating_segments_.contains(candidate)) {
```
Comment (MBkkt) [inline core/index/index_writer.cpp:1470]: strange. Is nullptr contains in map?

File: core/index/index_writer.cpp [hunk 2]
```diff
@@ -1503,10 +1510,10 @@ index_writer::consolidation_result index_writer::consolidate(
   {
     size_t found = 0;
 
-    const auto not_found = candidates.end();
     for (const auto& segment : *committed_meta) {
-      const auto it = std::lower_bound(candidates.begin(), not_found, &segment.meta);
-      found += (it != not_found && *it == &segment.meta);
+      const auto it = std::lower_bound(std::begin(candidates),
+                                       std::end(candidates), &segment.meta);
+      found += (it != std::end(candidates) && *it == &segment.meta);
```
Comment (MBkkt) [inline core/index/index_writer.cpp:1516]: It's ok, but why?

File: core/index/merge_writer.cpp [hunk 3]
```diff
@@ -162,10 +162,8 @@ class remapping_doc_iterator final : public doc_iterator {
  public:
   remapping_doc_iterator(doc_iterator::ptr&& it,
                          const doc_map_f& mapper) noexcept
-      : it_{std::move(it)}, mapper_{&mapper} {
-    assert(it_);
-    src_ = irs::get<document>(*it_);
-    assert(src_);
+      : it_{std::move(it)}, mapper_{&mapper}, src_{irs::get<document>(*it_)} {
```
Comment (MBkkt) [inline core/index/merge_writer.cpp:165]: IMO was better

File: core/index/merge_writer.cpp [hunk 4]
```diff
@@ -1556,16 +1568,22 @@ bool merge_writer::flush_sorted(tracking_directory& dir,
...
-    assert(it.payload);
-    auto& payload = it.payload->value;
+  for (doc_id_t next_id = doc_limits::min(); columns_it.next();) {
+    const auto [index, it] = columns_it.value();
+    assert(it);
+
+    if (IRS_UNLIKELY(!it->valid())) {
+      assert(false);
+      IR_FRMT_ERROR(
+          "Get an invalid iterator during consolidationg of sorted index, "
```
Comment (Dronplane) [inline core/index/merge_writer.cpp:1578]:
```suggestion
          "Got an invalid iterator during consolidationg of sorted index, "
```

File: core/index/merge_writer.cpp [hunk 5]
```diff
@@ -1556,16 +1568,22 @@ bool merge_writer::flush_sorted(tracking_directory& dir,
...
-    const auto value = columns_it.value();
-    assert(value.second);
-    auto& it = *value.second;
-    assert(it.payload);
-    auto& payload = it.payload->value;
+  for (doc_id_t next_id = doc_limits::min(); columns_it.next();) {
+    const auto [index, it] = columns_it.value();
+    assert(it);
+
+    if (IRS_UNLIKELY(!it->valid())) {
```
Comment (Dronplane) [inline core/index/merge_writer.cpp:1575]:
```suggestion
    if (IRS_UNLIKELY(!it || !it->valid())) {
```

**Review by MBkkt — APPROVED** [review]

File: core/index/merge_writer.cpp [hunk 6]
```diff
@@ -1132,10 +1128,14 @@ bool write_columns(columnstore& cs, Iterator& columns,
...
       auto it = column.iterator(ColumnHint::kConsolidation);
-      assert(it);
 
-      if (IRS_LIKELY(it)) {
+      if (IRS_LIKELY(it && irs::get<document>(*it))) {
         itrs.emplace_back(std::move(it), doc_map);
+      } else {
+        assert(false);
+        IR_FRMT_ERROR(
+            "Get an invalid iterator during consolidationg of the columnstore, "
```
Comment (Dronplane) [inline core/index/merge_writer.cpp:1137]:
```suggestion
            "Got an invalid iterator during consolidationg of the columnstore, "
```

**Review by Dronplane — APPROVED** [review]
LGTM. Some minor comments

File: core/index/merge_writer.cpp (hunk 3 above)
Comment (gnusi) [reply core/index/merge_writer.cpp:165]: Not it's checked outside

**Pull request merged**